### PR TITLE
[MCCAS][nfc] Refactor Dwarf Section handling

### PR DESCRIPTION
**Pull request by felipepiovezan**
Now that there is more clarity on how Dwarf sections are going to be
handled, this commit refactors the existing code into a more uniform
approach for handling these sections.

1. All the queries to the MCAssembler that attempt to find debug
information are only done once and stored into a cache object for ease
of access.

2. The main loop that iterates over all sections now dispatches the
handling of debug sections to specialized functions, removing any
handling that was previously inlined into the loop.

3. The specialized functions above now have a more consistent naming and
documentation schemes.

This is not the full extent of possible refactors, but it provides a
better starting point for the upcoming work, which will split more
complicated sections.

### CODE REVIEW

**Comment by felipepiovezan** [comment]
Rebased and fixed some typos.

File: llvm/lib/MC/MCCASObjectV1.cpp [hunk 1]
```diff
@@ -1209,18 +1219,13 @@ void MCCASBuilder::startSection(const MCSection *Sec) {
...
-  auto DwarfLineSection =
-      Asm.getContext().getObjectFileInfo()->getDwarfLineSection();
 
   if (RelocLocation == Atom) {
     // Build a map for lookup.
     for (auto R : ObjectWriter.getRelocations()[Sec]) {
       // For the Dwarf Sections, just append the relocations to the
       // SectionRelocs. No Atoms are considered for this section.
-      if (R.F && DwarfLineSection != Sec && Sec != DebugInfoSection)
+      if (R.F && DwarfSections.Line != Sec && Sec != DwarfSections.DebugInfo)
```
Comment (rastogishubham) [inline llvm/lib/MC/MCCASObjectV1.cpp]: Small nitpick, but can you change this to either be 
`if (R.F && Sec != DwarfSections.Line && Sec != DwarfSections.DebugInfo)` or `if (R.F && DwarfSections.Line != Sec && DwarfSections.DebugInfo != Sec)` so it looks consistent?

**Review by rastogishubham — APPROVED** [review]
LGTM otherwise! Small nitpick aside!

**Comment by rastogishubham** [comment]
@swift-ci please test

**Review by cachemeifyoucan — APPROVED** [review]

**Comment by felipepiovezan** [comment]
Rebased, addressed review comments

**Comment by felipepiovezan** [comment]
@swift-ci please test

**Pull request merged**